pkg/kubelet/nodeshutdown/nodeshutdown_manager_linux.go

```diff
@@ -395,19 +400,45 @@ func (m *managerImpl) processShutdownEvent() error {
 			}(pod, group)
 		}
 
+		// This duration determines how long the shutdown manager will wait for the pods in this group
+		// to terminate before proceeding to the next group.
+		var groupTerminationWaitDuration = time.Duration(group.ShutdownGracePeriodSeconds) * time.Second
 		var (
-			doneCh = make(chan struct{})
-			timer  = m.clock.NewTimer(time.Duration(group.ShutdownGracePeriodSeconds) * time.Second)
+			doneCh         = make(chan struct{})
+			timer          = m.clock.NewTimer(groupTerminationWaitDuration)
+			ctx, ctxCancel = context.WithTimeout(context.Background(), groupTerminationWaitDuration)
 		)
 		go func() {
 			defer close(doneCh)
+			defer ctxCancel()
 			wg.Wait()
+
+			// The signal to kill a Pod was sent successfully to all the pods,
+			// let's wait until all the volumes are unmounted from all the pods before
+			// continuing to the next group. This is done so that the CSI Driver (assuming
+			// that it's part of the highest group) has a chance to perform unmounts.
+			if err := m.volumeManager.WaitForAllPodsUnmount(ctx, group.Pods); err != nil {
```

Comment:
if timed out - will it exit with error? If not - should we somehow have a log statement about it?

Reply:
If any of the `WaitForUnmount` calls time out and return an error (which happens when volumes are still mounted after the timeout), those errors will be included in the aggregate error, which is already logged.